ppx.ml

```diff
@@ -0,0 +1,133 @@
+module Stre = Str
+open Parsetree
+open Ast_helper
+module AM = Ast_mapper
+
+let error ppf =
+  Format.kasprintf (fun s ->
+    let loc = !default_loc in
+    Exp.extension ~loc @@
+    AM.extension_of_error (Location.error ~loc s))
+    ppf
+
+let mklid m lid =
+  Exp.ident
+    {Location.loc = !default_loc ; txt = Longident.(Ldot (Lident m, lid))}
+let app m lid l =
+  Exp.(apply (mklid m lid)
+      (List.map (fun e -> Asttypes.Nolabel, e) l))
+
+let try_int s =
+  try `Int32 (Int32.of_string s)
+  with _ ->
+  try `Int64 (Int64.of_string s)
+  with _ -> `ZInt s
+
+let integer m s =
+  match try_int s with
+  | `Int32 0l -> mklid m "zero"
+  | `Int32 1l -> mklid m "one"
+  | `Int32 (-1l) -> mklid m "minus_one"
+  | `Int32 i -> app m "of_int32" [Exp.constant @@ Const.int32 i]
+  | `Int64 i -> app m "of_int64" [Exp.constant @@ Const.int64 i]
+  | `ZInt s -> app m "of_string" [Exp.constant @@ Const.string s]
+
+let integer_z = integer "Z"
+let integer_q = integer "Q"
+
+(* float_of_string doesn't error on floats that are not representable so
```

Comment:
Sorry to comment on a comment, but as phrased this confused me more than it helped me.

How about replacing the sentence “float_of_string doesn't error on floats …” by:
```
When the programmer writes 9007199254740993.0q, they want the rational for 9007199254740993, not for the closest double-precision number to that.
Similarly, when the programmer writes 0.1q, they want the rational for 1/10, not for 1000000000000000055511151231257827021181583404541015625/100000000000000000000000000000000000000000000000000000000.
For all these reasons float_of_string must not be used here.
```